fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,784 @@
+
+                // Use monoid to combine value for the given key
+                // if the memory and disk have diverged
+                if existing_value_updated {
+                    let resolved_value = V::combine(existing_value, new_value);
+                    self.data.insert(key.clone(), resolved_value);
+                } else {
+                    self.data.insert(key.clone(), new_value.clone());
+                }
+            } else {
+                self.data.insert(key.clone(), new_value.clone());
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Remove files from disk that are not present in memory
+    fn remove_files_not_in_ram(&mut self) -> Result<()> {
+        for key in self.deleted_keys.iter() {
+            let file_path = self.path.join(format!("{}.json", key));
+            if file_path.exists() {
+                fs::remove_file(&file_path).expect("Failed to delete file");
+            }
+        }
+        Ok(())
+    }
+
+    pub fn extract_key_from_file_path(&self, path: &Path) -> Result<K> {
+        path.file_stem()
+            .ok_or_else(|| {
+                ArklibError::Storage(
+                    self.label.clone(),
+                    "Failed to extract file stem from filename".to_owned(),
+                )
+            })?
+            .to_str()
+            .ok_or_else(|| {
+                ArklibError::Storage(
+                    self.label.clone(),
+                    "Failed to convert file stem to string".to_owned(),
+                )
+            })?
+            .parse::<K>()
+            .map_err(|_| {
+                ArklibError::Storage(
+                    self.label.clone(),
+                    "Failed to parse key from filename".to_owned(),
+                )
+            })
+    }
+}
+
+impl<K, V> BaseStorage<K, V> for FolderStorage<K, V>
+where
+    K: Ord
+        + Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr
+        + std::fmt::Display,
+    V: Clone + serde::Serialize + serde::de::DeserializeOwned + Monoid<V>,
+{
+    /// Set a key-value pair in the internal mapping
+    fn set(&mut self, key: K, value: V) {
+        self.data.insert(key.clone(), value);
+        self.deleted_keys.remove(&key);
+        self.modified.insert(key, SystemTime::now());
+    }
+
+    /// Remove an entry from the internal mapping given a key
+    fn remove(&mut self, id: &K) -> Result<()> {
+        match self.data.remove(id) {
+            Some(_) => {
+                self.deleted_keys.insert(id.clone());
+                Ok(())
+            }
+            None => Err(ArklibError::Storage(
+                self.label.clone(),
+                "Key not found".to_owned(),
+            )),
+        }
+    }
+
+    /// Compare the timestamp of the storage files
+    /// with the timestamps of the in-memory storage and the last written
+    /// to time to determine if either of the two requires syncing.
```

Comment:
Sorry, I don't fully understand this sentence.

Reply:
updated it:
```rs
    /// Determines the synchronization status between RAM and disk.
    /// Compares modification timestamps of files in RAM and on disk,
    /// checking for newer versions in path.

```